utilities/ovs-vlan-bug-workaround.8.in
.\" -*- nroff -*-
.so lib/ovs.tmac
.TH ovs\-vlan\-bug\-workaround 8 "@VERSION@" "Open vSwitch" "Open vSwitch Manual"
.ds PN ovs\-vlan\-bug\-workaround
.
.SH NAME
ovs\-vlan\-bug\-workaround \- utility for configuring Linux VLAN driver bug workaround
.
.SH SYNOPSIS
\fBovs\-vlan\-bug\-workaround \fInetdev\fR \fBon\fR
.br
\fBovs\-vlan\-bug\-workaround \fInetdev\fR \fBoff\fR
.br
\fBovs\-vlan\-bug\-workaround \-\-help
.br
\fBovs\-vlan\-bug\-workaround \-\-version
.SH DESCRIPTION
.
.PP
Some Linux network drivers support a feature called ``VLAN
acceleration''.  VLAN acceleration is associated with a data structure
called a \fBvlan_group\fR that is, abstractly, a dictionary that maps
from a VLAN ID (in the range 0 to 4095) to a VLAN device, that is, a
Linux network device associated with a particular VLAN,
e.g. \fBeth0.9\fR for VLAN 9 on \fBeth0\fR.
.PP
Some drivers that support VLAN acceleration have bugs that fall
roughly into the categories listed below.  \fBovs\-vlan\-test\fR(8)
can test for these driver bugs.
.so utilities/ovs-vlan-bugs.man
.PP
.PP
The correct long term solution is to fix these driver bugs.
.PP
For now, \fBovs\-vlan\-bug\-workaround\fR can enable a special-purpose
workaround for devices with buggy VLAN acceleration.  A kernel patch
must be applied for this workaround to work.
.PP
Use the command \fBovs\-vlan\-bug\-workaround \fInetdev\fR \fBon\fR to
enable the VLAN driver bug workaround for network device \fInetdev\fR.
Use the command \fBovs\-vlan\-bug\-workaround \fInetdev\fR \fBoff\fR to
disable the VLAN driver bug workaround for network device \fInetdev\fR.
.SH "DRIVER DETAILS"
.PP
The following drivers in Linux version
2.6.32.12-0.7.1.xs1.0.0.311.170586 implement VLAN acceleration and are
relevant to Open vSwitch on XenServer.  We have not tested any version
of most of these drivers, so we do not know whether they have a VLAN
problem that needs to be fixed.  The drivers are listed by the name
that they report in, e.g., \fBethtool \-i\fR output; in a few cases
this differs slightly from the name of the module's \fB.ko\fR file:
.
.nf
.ta T 1i
\fB8139cp	acenic	amd8111e	atl1c	ATL1E	atl1
atl2	be2net	bna	bnx2	bnx2x	cnic
cxgb	cxgb3	e1000	e1000e	enic	forcedeth
igb	igbvf	ixgb	ixgbe	jme	ml4x_core
ns83820	qlge	r8169	S2IO	sky2	starfire
tehuti	tg3	typhoon	via-velocity	vxge
.fi
.PP
The following drivers use \fBvlan_group\fR but are irrelevant to Open
vSwitch on XenServer:
.IP "\fBbonding\fR"
Not used with Open vSwitch on XenServer.
.IP "\fBgianfar\fR"
Not shipped with XenServer.  A FreeScale CPU-integrated device.
.IP "\fBehea\fR"
Cannot be built on x86.  IBM Power architecture only.
.IP "\fBstmmac\fR"
Cannot be built on x86.  SH4 architecture only.
.IP "\fBvmxnet3\fR"
Not shipped with XenServer.  For use inside VMware VMs only.
.
.SH OPTIONS
.
.so lib/common.man
.
.SH BUGS
.
Obviously.
.
.SH "SEE ALSO"
.
.BR ovs\-vlan\-test (8).


NEWS
v2.5.7 - xx xxx xxxx
--------------------

v2.5.6 - 19 Oct 2018
--------------------
   - Bug fixes
   - Bug fixes

v2.5.5 - 18 May 2018
---------------------
   - Bug fixes

v2.5.4 - 03 Sep 2017
---------------------
   - Bug fixes

v2.5.3 - 28 Jun 2017
---------------------
   - Bug fixes

v2.5.2 - 21 Feb 2017
---------------------
   - ovs-ctl:
     * Added the ability to selectively start the forwarding and database
       functions (ovs-vswitchd and ovsdb-server, respectively).
   - Fedora packaging:
     * A package upgrade does not automatically restart OVS service.
   - Bug fixes

v2.5.1 - 27 Sep 2016
---------------------
   - DPDK:
     * New appctl command 'dpif-netdev/pmd-rxq-show' to check the port/rxq
       assignment.
     * Type of log messages from PMD threads changed from INFO to DBG.
   - ovs-pki: Changed message digest algorithm from SHA-1 to SHA-512 because
     SHA-1 is no longer secure and some operating systems have started to
     disable it in OpenSSL.
   - Bug fixes

v2.5.0 - 26 Feb 2016
---------------------
   - Dropped support for Python older than version 2.7.  As a consequence,
     using Open vSwitch 2.5 or later on XenServer 6.5 or earlier (which
     have Python 2.4) requires first installing Python 2.7.
   - OpenFlow:
     * Group chaining (where one OpenFlow group triggers another) is
       now supported.
     * OpenFlow 1.4+ "importance" is now considered for flow eviction.
     * OpenFlow 1.4+ OFPTC_EVICTION is now implemented.
     * OpenFlow 1.4+ OFPTC_VACANCY_EVENTS is now implemented.
     * OpenFlow 1.4+ OFPMP_TABLE_DESC is now implemented.
     * Allow modifying the ICMPv4/ICMPv6 type and code fields.
     * OpenFlow 1.4+ OFPT_SET_ASYNC_CONFIG and OFPT_GET_ASYNC_CONFIG are
       now implemented.
   - ovs-ofctl:
     * New "out_group" keyword for OpenFlow 1.1+ matching on output group.
   - Tunnels:
     * Geneve tunnels can now match and set options and the OAM bit.
     * The nonstandard GRE64 tunnel extension has been dropped.
   - Support Multicast Listener Discovery (MLDv1 and MLDv2).
   - Add 'symmetric_l3l4' and 'symmetric_l3l4+udp' hash functions.
   - sFlow agent now reports tunnel and MPLS structures.
   - New 'check-system-userspace', 'check-kmod' and 'check-kernel' Makefile
     targets to run a new system testsuite.  These tests can be run inside
     a Vagrant box.  See INSTALL.md for details
   - Mark --syslog-target argument as deprecated.  It will be removed in
     the next OVS release.
   - Added --user option to all daemons
   - Add support for connection tracking through the new "ct" action
     and "ct_state"/"ct_zone"/"ct_mark"/"ct_label" match fields.  Only
     available on Linux kernels with the connection tracking module loaded.
   - Add experimental version of OVN.  OVN, the Open Virtual Network, is a
     system to support virtual network abstraction.  OVN complements the
     existing capabilities of OVS to add native support for virtual network
     abstractions, such as virtual L2 and L3 overlays and security groups.
   - RHEL packaging:
     * DPDK ports may now be created via network scripts (see README.RHEL).
   - DPDK:
     * Requires DPDK 2.2
     * Added multiqueue support to vhost-user
     * Note: QEMU 2.5+ required for multiqueue support
   - SELinux:
     * Introduced SELinux policy package.

v2.4.0 - 20 Aug 2015
---------------------
   - Flow table modifications are now atomic, meaning that each packet
     now sees a coherent version of the OpenFlow pipeline.  For
     example, if a controller removes all flows with a single OpenFlow
     "flow_mod", no packet sees an intermediate version of the OpenFlow
     pipeline where only some of the flows have been deleted.
   - Added support for SFQ, FQ_CoDel and CoDel qdiscs.
   - Add bash command-line completion support for ovs-vsctl Please check
     utilities/ovs-command-compgen.INSTALL.md for how to use.
   - The MAC learning feature now includes per-port fairness to mitigate
     MAC flooding attacks.
   - New support for a "conjunctive match" OpenFlow extension, which
     allows constructing OpenFlow matches of the form "field1 in
     {a,b,c...} AND field2 in {d,e,f...}" and generalizations.  For details,
     see documentation for the "conjunction" action in ovs-ofctl(8).
   - Add bash command-line completion support for ovs-appctl/ovs-dpctl/
     ovs-ofctl/ovsdb-tool commands.  Please check
     utilities/ovs-command-compgen.INSTALL.md for how to use.
   - The "learn" action supports a new flag "delete_learned" that causes
     the learned flows to be deleted when the flow with the "learn" action
     is deleted.
   - Basic support for the Geneve tunneling protocol. It is not yet
     possible to generate or match options. This is planned for a future
     release. The protocol is documented at
     http://tools.ietf.org/html/draft-gross-geneve-00
   - The OVS database now reports controller rate limiting statistics.
   - sflow now exports information about LACP-based bonds, port names, and
     OpenFlow port numbers, as well as datapath performance counters.
   - ovs-dpctl functionality is now available for datapaths integrated
     into ovs-vswitchd, via ovs-appctl.  Some existing ovs-appctl
     commands are now redundant and will be removed in a future
     release.  See ovs-vswitchd(8) for details.
   - OpenFlow:
     * OpenFlow 1.4 bundles are now supported for flow mods and port
       mods.  For flow mods, both 'atomic' and 'ordered' bundle flags
       are trivially supported, as all bundled messages are executed
       in the order they were added and all flow table modifications
       are now atomic to the datapath.  Port mods may not appear in
       atomic bundles, as port status modifications are not atomic.
     * IPv6 flow label and neighbor discovery fields are now modifiable.
     * OpenFlow 1.5 extended registers are now supported.
     * The OpenFlow 1.5 actset_output field is now supported.

utilities/ovs-vlan-bug-workaround.c
/*
 * Copyright (c) 2011 Nicira, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at:
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <config.h>

#include <errno.h>
#include <getopt.h>
#include <linux/if_vlan.h>
#include <linux/sockios.h>
#include <net/if.h>
#include <stdlib.h>
#include <sys/ioctl.h>
#include <sys/socket.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>

#include "command-line.h"
#include "util.h"

#define ADD_ALL_VLANS_CMD 10
#define DEL_ALL_VLANS_CMD 11

static void usage(void);
static void parse_options(int argc, char *argv[]);

int
main(int argc, char *argv[])
{
    struct vlan_ioctl_args vlan_args;
    const char *netdev, *setting;
    int fd;

    set_program_name(argv[0]);

    parse_options(argc, argv);
    if (argc - optind != 2) {
        ovs_fatal(0, "exactly two non-option arguments are required "
                  "(use --help for help)");
    }

    memset(&vlan_args, 0, sizeof vlan_args);

    /* Get command. */
    setting = argv[optind + 1];
    if (!strcmp(setting, "on")) {
        vlan_args.cmd = ADD_ALL_VLANS_CMD;
    } else if (!strcmp(setting, "off")) {
        vlan_args.cmd = DEL_ALL_VLANS_CMD;
    } else {
        ovs_fatal(0, "second command line argument must be \"on\" or \"off\" "
                  "(not \"%s\")", setting);
    }

    /* Get network device name. */
    netdev = argv[optind];
    if (strlen(netdev) >= IFNAMSIZ) {
        ovs_fatal(0, "%s: network device name too long", netdev);
    }
    strcpy(vlan_args.device1, netdev);

    /* Execute operation. */
    fd = socket(AF_INET, SOCK_STREAM, 0);
    if (fd < 0) {
        ovs_fatal(errno, "socket creation failed");
    }
    if (ioctl(fd, SIOCSIFVLAN, &vlan_args) < 0) {
        if (errno == ENOPKG) {
            ovs_fatal(0, "operation failed (8021q module not loaded)");
        } else if (errno == EOPNOTSUPP) {
            ovs_fatal(0, "operation failed (kernel does not support the "
                      "VLAN bug workaround)");
        } else {
            ovs_fatal(errno, "operation failed");
        }
    }
    close(fd);

    return 0;
}

static void
usage(void)
{
    printf("\
%s, for enabling or disabling the kernel VLAN bug workaround\n\
usage: %s NETDEV SETTING\n\
where NETDEV is a network device (e.g. \"eth0\")\n\
  and SETTING is \"on\" to enable the workaround or \"off\" to disable it.\n\
\n\
Options:\n\
  -h, --help         Print this helpful information\n\
  -V, --version      Display version information\n",
           program_name, program_name);
    exit(EXIT_SUCCESS);
}

static void
parse_options(int argc, char *argv[])
{
    static const struct option long_options[] = {
        {"help", no_argument, NULL, 'h'},
        {"version", no_argument, NULL, 'V'},
        {NULL, 0, NULL, 0},
    };
    char *short_options = ovs_cmdl_long_options_to_short_options(long_options);

    for (;;) {
        int option;

        option = getopt_long(argc, argv, "+t:hVe", long_options, NULL);
        if (option == -1) {
            break;
        }
        switch (option) {
        case 'h':
            usage();
            break;

        case 'V':
            ovs_print_version(0, 0);
            exit(EXIT_SUCCESS);

        case '?':
            exit(EXIT_FAILURE);

        default:
            OVS_NOT_REACHED();
        }
    }
    free(short_options);
}
